src/Bicep.Core/Diagnostics/DiagnosticBuilder.cs

```diff
@@ -1491,6 +1491,21 @@ public ErrorDiagnostic UnknownModuleReferenceScheme(string badScheme, ImmutableA
                 TextSpan,
                 "BCP254",
                 "Expected a Bicep file path string. This should be a relative path to another bicep file, e.g. 'myModule.bicep' or '../parent/myModule.bicep'");
+
+            public ErrorDiagnostic MissingParameterAssignment(string identifier) => new(
+                TextSpan,
+                "BCP255",
+                $"\"{identifier}\" is declared in the bicep file but is missing an assignment in the params file.");
+
+            public ErrorDiagnostic MissingParameterDeclaration(string? identifier) => new(
+                TextSpan,
+                "BCP256",
+                $"\"{identifier}\" is assigned in the params file without being declared in the bicep file.");
+
+            public ErrorDiagnostic ParameterTypeMismatch(string? identifier, TypeSymbol expectedType, TypeSymbol actualType) => new(
+                TextSpan,
+                "BCP257",
+                $"The parameter \"{identifier}\" expects a value of type \"{expectedType}\" but the provided value is of type \"{actualType}\".");
```

Comment:
[nit] I like the way you start the message with `The parameter "..."` here - would be good to update BCP255 & BCP256 to be consistent with this.